### ISSUE TITLE
UnsatisfiedLinkError of libjnitensorflow.dylib

### ISSUE THREAD

**Issue by sj6077** [bytedeco/javacpp-presets#335]
There is a linking error in tensorflow. I used maven and I added below to the pom.xml
```
<dependency>
      <groupId>org.bytedeco.javacpp-presets</groupId>
      <artifactId>tensorflow-platform</artifactId>
      <version>0.11.0-1.3</version>
</dependency>
```

My platform is macosx-x86_64. Is there anything that I missed?
Full error messages are here.

```
java.lang.UnsatisfiedLinkError: no jnitensorflow in java.library.path
at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1867)
at java.lang.Runtime.loadLibrary0(Runtime.java:870)
at java.lang.System.loadLibrary(System.java:1122)
at org.bytedeco.javacpp.Loader.loadLibrary(Loader.java:804)
at org.bytedeco.javacpp.Loader.load(Loader.java:613)
at org.bytedeco.javacpp.Loader.load(Loader.java:530)
at org.bytedeco.javacpp.tensorflow.(tensorflow.java:10)
at java.lang.Class.forName0(Native Method)
at java.lang.Class.forName(Class.java:348)
at org.bytedeco.javacpp.Loader.load(Loader.java:585)
at org.bytedeco.javacpp.Loader.load(Loader.java:530)
at org.bytedeco.javacpp.tensorflow$Scope.(tensorflow.java:14559)
at edu.snu.cay.dolphin.async.tensorflow.sample_example.TensorflowTrainer.initialize(TensorflowTrainer.java:60)
at edu.snu.cay.dolphin.async.AsyncWorkerTask.call(AsyncWorkerTask.java:75)
at org.apache.reef.runtime.common.evaluator.task.TaskRuntime.runTask(TaskRuntime.java:277)
at org.apache.reef.runtime.common.evaluator.task.TaskRuntime.run(TaskRuntime.java:137)
at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.UnsatisfiedLinkError: /Users/soo/.javacpp/cache/dolphin-async-0.1-SNAPSHOT-shaded.jar/org/bytedeco/javacpp/macosx-x86_64/libjnitensorflow.dylib: dlopen(/Users/soo/.javacpp/cache/dolphin-async-0.1-SNAPSHOT-shaded.jar/org/bytedeco/javacpp/macosx-x86_64/libjnitensorflow.dylib, 1): Symbol not found: __ZN10tensorflow19kColocationAttrNameE
Referenced from: /Users/soo/.javacpp/cache/dolphin-async-0.1-SNAPSHOT-shaded.jar/org/bytedeco/javacpp/macosx-x86_64/libjnitensorflow.dylib
Expected in: /Users/soo/.javacpp/cache/dolphin-async-0.1-SNAPSHOT-shaded.jar/org/bytedeco/javacpp/macosx-x86_64/./libtensorflow.so
in /Users/soo/.javacpp/cache/dolphin-async-0.1-SNAPSHOT-shaded.jar/org/bytedeco/javacpp/macosx-x86_64/libjnitensorflow.dylib
at java.lang.ClassLoader$NativeLibrary.load(Native Method)
at java.lang.ClassLoader.loadLibrary0(ClassLoader.java:1941)
at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1824)
at java.lang.Runtime.load0(Runtime.java:809)
at java.lang.System.load(System.java:1086)
at org.bytedeco.javacpp.Loader.loadLibrary(Loader.java:784)
... 13 more

```

**Comment by sj6077** [comment]
My platform is macosx-x86_64.

Full error messages are here.

java.lang.UnsatisfiedLinkError: no jnitensorflow in java.library.path
	at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1867)
	at java.lang.Runtime.loadLibrary0(Runtime.java:870)
	at java.lang.System.loadLibrary(System.java:1122)
	at org.bytedeco.javacpp.Loader.loadLibrary(Loader.java:804)
	at org.bytedeco.javacpp.Loader.load(Loader.java:613)
	at org.bytedeco.javacpp.Loader.load(Loader.java:530)
	at org.bytedeco.javacpp.tensorflow.<clinit>(tensorflow.java:10)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at org.bytedeco.javacpp.Loader.load(Loader.java:585)
	at org.bytedeco.javacpp.Loader.load(Loader.java:530)
	at org.bytedeco.javacpp.tensorflow$Scope.<clinit>(tensorflow.java:14559)
	at edu.snu.cay.dolphin.async.tensorflow.sample_example.TensorflowTrainer.initialize(TensorflowTrainer.java:60)
	at edu.snu.cay.dolphin.async.AsyncWorkerTask.call(AsyncWorkerTask.java:75)
	at org.apache.reef.runtime.common.evaluator.task.TaskRuntime.runTask(TaskRuntime.java:277)
	at org.apache.reef.runtime.common.evaluator.task.TaskRuntime.run(TaskRuntime.java:137)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.UnsatisfiedLinkError: /Users/soo/.javacpp/cache/dolphin-async-0.1-SNAPSHOT-shaded.jar/org/bytedeco/javacpp/macosx-x86_64/libjnitensorflow.dylib: dlopen(/Users/soo/.javacpp/cache/dolphin-async-0.1-SNAPSHOT-shaded.jar/org/bytedeco/javacpp/macosx-x86_64/libjnitensorflow.dylib, 1): Symbol not found: __ZN10tensorflow19kColocationAttrNameE
  Referenced from: /Users/soo/.javacpp/cache/dolphin-async-0.1-SNAPSHOT-shaded.jar/org/bytedeco/javacpp/macosx-x86_64/libjnitensorflow.dylib
  Expected in: /Users/soo/.javacpp/cache/dolphin-async-0.1-SNAPSHOT-shaded.jar/org/bytedeco/javacpp/macosx-x86_64/./libtensorflow.so
 in /Users/soo/.javacpp/cache/dolphin-async-0.1-SNAPSHOT-shaded.jar/org/bytedeco/javacpp/macosx-x86_64/libjnitensorflow.dylib
	at java.lang.ClassLoader$NativeLibrary.load(Native Method)
	at java.lang.ClassLoader.loadLibrary0(ClassLoader.java:1941)
	at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1824)
	at java.lang.Runtime.load0(Runtime.java:809)
	at java.lang.System.load(System.java:1086)
	at org.bytedeco.javacpp.Loader.loadLibrary(Loader.java:784)
	... 13 more

**Comment by saudet** [comment]
It works fine here on Yosemite. What version of Mac OS X are you using?

**Comment by sj6077** [comment]
It's Mac OS X v10.11.6. According to the cppbuild.sh of tensorflow, only gpu version is supported, is it right? My computer doesn't have any GPU, so I build tensorflow cpu version. I wonder it is the reason or not.

**Comment by saudet** [comment]
No it's not related. I'm pretty sure it's just because your shaded JAR files contains the wrong version of the library.

**Comment by sj6077** [comment]
I built it as a same way on ubuntu, it works. I'm not sure because I installed tensorflow v0.11 using pip install.

**Comment by saudet** [comment]
Yes, you have an old version somewhere. `libtensorflow.so` doesn't come from `tensorflow-0.11.0-1.3-macosx-x86_64`, that comes with `libtensorflow_cc.so`. You'll need to find where `libtensorflow.so` comes from and get rid of that.

**Comment by ByzenMa** [comment]
Hi, @saudet , I got a same linking error in tensorflow when  I used version 0.11.0-1.3 in maven. Full error messages are as follows:
Exception in thread "main" java.lang.UnsatisfiedLinkError: no jnitensorflow in java.library.path
	at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1886)
	at java.lang.Runtime.loadLibrary0(Runtime.java:849)
	at java.lang.System.loadLibrary(System.java:1088)
	at org.bytedeco.javacpp.Loader.loadLibrary(Loader.java:804)
	at org.bytedeco.javacpp.Loader.load(Loader.java:613)
	at org.bytedeco.javacpp.Loader.load(Loader.java:530)
	at org.bytedeco.javacpp.tensorflow.<clinit>(tensorflow.java:10)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:274)
	at org.bytedeco.javacpp.Loader.load(Loader.java:585)
	at org.bytedeco.javacpp.Loader.load(Loader.java:530)
	at org.bytedeco.javacpp.helper.tensorflow$AbstractSession.<clinit>(tensorflow.java:214)
	at com.chanjet.ai.nlp.s2s.datahandler.TensorflowExample.<init>(TensorflowExample.java:23)
	at com.chanjet.ai.nlp.s2s.datahandler.TensorflowExample.main(TensorflowExample.java:68)
Caused by: java.lang.UnsatisfiedLinkError: /Users/mabaizhang/.javacpp/cache/tensorflow-0.11.0-1.3-macosx-x86_64.jar/org/bytedeco/javacpp/macosx-x86_64/libjnitensorflow.dylib: dlopen(/Users/mabaizhang/.javacpp/cache/tensorflow-0.11.0-1.3-macosx-x86_64.jar/org/bytedeco/javacpp/macosx-x86_64/libjnitensorflow.dylib, 1): Library not loaded: @rpath/libtensorflow.so
  Referenced from: /Users/mabaizhang/.javacpp/cache/tensorflow-0.11.0-1.3-macosx-x86_64.jar/org/bytedeco/javacpp/macosx-x86_64/libjnitensorflow.dylib
  Reason: image not found
	at java.lang.ClassLoader$NativeLibrary.load(Native Method)
	at java.lang.ClassLoader.loadLibrary1(ClassLoader.java:1965)
	at java.lang.ClassLoader.loadLibrary0(ClassLoader.java:1890)
	at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1851)
	at java.lang.Runtime.load0(Runtime.java:795)
	at java.lang.System.load(System.java:1062)
	at org.bytedeco.javacpp.Loader.loadLibrary(Loader.java:784)
	... 10 more
Is there anything wrong with my tensorflow version somewhere?  My platform is macosx-x86_64

**Comment by saudet** [comment]
I can't reproduce that on my system, but the install name of the library is
obviously wrong. This is a bug in TensorFlow, so please report upstream.
Thank you!

**Comment by ByzenMa** [comment]
OK, thanks for replying!

**Comment by lefromage** [comment]
is there a workaround ?

**Comment by saudet** [comment]
@lefromage Yes, I included a workaround here until they decide to fix this upstream somewhere:
https://github.com/bytedeco/javacpp-presets/commit/fa4fe01d5931c6beaf76bc72eb76e407b2ecb181#diff-87f133fa4ea1f42529d4a75dfbeb247eR79

**Comment by lefromage** [comment]
Thanks @saudet 

for those still using 0.11.0-1.3 the workaround on mac or linux 
is to build libtensorflow.so from source and refer to it's location in LD_LIBRARY_PATH

git clone  -b v0.11.0 --recurse-submodules https://github.com/tensorflow/tensorflow tensorflow-v0.11.0
cd tensorflow-v0.11.0
# edit and replace in ./tensorflow/workspace.bzl , zlib-1.2.8 code block with zlib-1.2.11
  native.new_http_archive(
    name = "zlib_archive",
    url = "http://zlib.net/zlib-1.2.11.tar.gz",
    sha256 = "c3e5e9fdd5004dcb542feda5ee4f0ff0744628baf8ed2dd5d66f8ca1197cb1a1",
    strip_prefix = "zlib-1.2.11",
    build_file = str(Label("//:zlib.BUILD")),
  )

./configure
bazel build //tensorflow:libtensorflow.so
export LD_LIBRARY_PATH=${PWD}/bazel-bin/tensorflow/libtensorflow.so:$LD_LIBRARY_PATH

**Comment by saudet** [comment]
Just renaming `libtensorflow_cc.so` to `libtensorflow.so` should also work...

**Comment by alistairstead3408** [comment]
This is an issue I am having with mac, linux and windows :(

**Comment by saudet** [comment]
@alistairstead3408 Please use the pom.xml file provided here: 
https://github.com/bytedeco/javacpp-presets/tree/1.3/tensorflow#the-pomxml-build-file

**Comment by saudet** [comment]
Resolved in binaries just released for version 1.0.0:
https://github.com/bytedeco/javacpp-presets/tree/master/tensorflow#the-pomxml-build-file
http://repo1.maven.org/maven2/org/bytedeco/javacpp-presets/tensorflow/1.0.0-1.3/
